### PR TITLE
Don't unblock run-level-concurrency-blocked runs in the resolver

### LINKED ISSUE #37446: Flaky test `TestScheduleConcurrency`
https://github.com/go-gitea/gitea/actions/runs/24971277825/job/73114689583

```
=== TestScheduleConcurrency (tests/integration/actions_concurrency_test.go:1122)
--- FAIL: TestScheduleConcurrency (3.07s)
    testlogger.go:62: 2026/04/27 00:50:36 modules/storage/local.go:48:NewLocalStorage() [I] Creating new Local Storage at /home/runner/work/gitea/gitea/tests/gitea-lfs-meta
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 5.2ms @ auth/auth.go:284(auth.SignInPost)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 6.0ms @ setting/applications.go:36(setting.ApplicationsPost)
    testlogger.go:62: 2026/04/27 00:50:36 .../actions/notifier_helper.go:165:notify() [W] Ref of event "repository" is empty, will fall back to the default branch
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 201 Created in 544.1ms @ repo/repo.go:292(repo.Create)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 3.5ms @ auth/auth.go:284(auth.SignInPost)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 5.3ms @ setting/applications.go:36(setting.ApplicationsPost)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 3.3ms @ auth/auth.go:284(auth.SignInPost)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 303 See Other in 5.1ms @ setting/applications.go:36(setting.ApplicationsPost)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST  for test-mock:12345, 200 OK in 7.8ms @ <autogenerated>:1(actions.API.CreateRegistrationToken-fm)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST /api/actions/ping.v1.PingService/Ping for 127.0.0.1:47640, 200 OK in 0.1ms @ <autogenerated>:1(http.Handler.ServeHTTP-fm)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST /api/actions/runner.v1.RunnerService/Register for 127.0.0.1:47640, 200 OK in 6.1ms @ <autogenerated>:1(http.Handler.ServeHTTP-fm)
    testlogger.go:62: 2026/04/27 00:50:36 HTTPRequest [I] router: completed POST /api/internal/hook/pre-receive/user2/actions-concurrency for 127.0.0.1:0, 200 OK in 5.1ms @ private/hook_pre_receive.go:109(private.HookPreReceive)
    testlogger.go:62: 2026/04/27 00:50:37 HTTPRequest [I] router: completed POST /api/internal/hook/post-receive/user2/actions-concurrency for 127.0.0.1:0, 200 OK in 184.7ms @ private/hook_post_receive.go:33(private.HookPostReceive)
    testlogger.go:62: 2026/04/27 00:50:37 HTTPRequest [I] router: completed POST  for test-mock:12345, 201 Created in 393.6ms @ repo/file.go:439(repo.CreateFile)
    actions_runner_test.go:103: 
        	Error Trace:	/home/runner/work/gitea/gitea/tests/integration/actions_runner_test.go:103
        	            				/home/runner/work/gitea/gitea/tests/integration/actions_concurrency_test.go:1167
        	            				/home/runner/work/gitea/gitea/tests/integration/git_helper_for_declarative_test.go:134
        	            				/home/runner/work/gitea/gitea/tests/integration/actions_concurrency_test.go:1122
        	Error:      	Expected nil, but got: &runnerv1.Task{state:impl.MessageState{NoUnkeyedLiterals:pragma.NoUnkeyedLiterals{}, DoNotCompare:pragma.DoNotCompare{}, DoNotCopy:pragma.DoNotCopy{}, atomicMessageInfo:(*impl.MessageInfo)(nil)}, sizeCache:0, unknownFields:[]uint8(nil), Id:93, WorkflowPayload:[]uint8{0x6e, 0x61, 0x6d, 0x65, 0x3a, 0x20, 0x73, 0x63, 0x68, 0x65, 0x64, 0x75, 0x6c, 0x65, 0x2d, 0x63, 0x6f, 0x6e, 0x63, 0x75, 0x72, 0x72, 0x65, 0x6e, 0x63, 0x79, 0xa, 0x22, 0x6f, 0x6e, 0x22, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x70, 0x75, 0x73, 0x68, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x73, 0x63, 0x68, 0x65, 0x64, 0x75, 0x6c, 0x65, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x2d, 0x20, 0x63, 0x72, 0x6f, 0x6e, 0x3a, 0x20, 0x27, 0x40, 0x65, 0x76, 0x65, 0x72, 0x79, 0x20, 0x31, 0x6d, 0x27, 0xa, 0x6a, 0x6f, 0x62, 0x73, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x6a, 0x6f, 0x62, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x6e, 0x61, 0x6d, 0x65, 0x3a, 0x20, 0x6a, 0x6f, 0x62, 0xa, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x72, 0x75, 0x6e, 0x73, 0x2d, 0x6f, 0x6e, 0x3a, 0x20, 0x75, 0x62, 0x75, 0x6e, 0x74, 0x75, 0x2d, 0x6c, 0x61, 0x74, 0x65, 0x73, 0x74, 0xa, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x73, 0x74, 0x65, 0x70, 0x73, 0x3a, 0xa, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x2d, 0x20, 0x72, 0x75, 0x6e, 0x3a, 0x20, 0x65, 0x63, 0x68, 0x6f, 0x20, 0x27, 0x73, 0x63, 0x68, 0x65, 0x64, 0x75, 0x6c, 0x65, 0x20, 0x77, 0x6f, 0x72, 0x6b, 0x66, 0x6c, 0x6f, 0x77, 0x27, 0xa}, Context:(*structpb.Struct)(0x24c5c244a480), Secrets:map[string]string{"GITEA_TOKEN":"92b2d57831c1a6e2dfdf1a33f0fba5cfb3d578c9", "GITHUB_TOKEN":"92b2d57831c1a6e2dfdf1a33f0fba5cfb3d578c9"}, Machine:"", Needs:map[string]*runnerv1.TaskNeed(nil), Vars:map[string]string(nil)}
        	Test:       	TestScheduleConcurrency
        	Messages:   	a task is fetched
    testlogger.go:62: 2026/04/27 00:50:38 services/actions/cleanup.go:170:CleanupEphemeralRunnersByPickedTaskOfRepo() [I] Removed 0 runners
    testlogger.go:62: 2026/04/27 00:50:38 .../repository/delete.go:363:DeleteRepositoryDirectly() [E] remove log file "user2/actions-concurrency/5d/93.log.zst": dbfs.Remove "actions_log/user2/actions-concurrency/5d/93.log.zst": file does not exist
    testlogger.go:62: 2026/04/27 00:50:38 .../repository/delete.go:363:DeleteRepositoryDirectly() [E] remove log file "user2/actions-concurrency/5c/92.log.zst": dbfs.Remove "actions_log/user2/actions-concurrency/5c/92.log.zst": file does not exist
    testlogger.go:62: 2026/04/27 00:50:38 HTTPRequest [I] router: completed DELETE  for test-mock:12345, 204 No Content in 36.8ms @ repo/repo.go:1073(repo.Delete)

### PR DESCRIPTION
**Pull request by silverwind**
Fixes #37446.

The job-status resolver in `checkJobsOfCurrentRunAttempt` only considered `needs` and job-level concurrency when transitioning jobs out of `Blocked`. When something drove the resolver against a run blocked solely by workflow-level concurrency — for example, a sibling run in the same group entering the queue and triggering `EmitJobsIfReadyByRun` — the run's job silently became `Waiting` while another run still held the concurrency group, and the runner could pick it up, defeating the concurrency guarantee.

The fix bails out of the resolver when the run's latest attempt is still blocked by run-level concurrency. `checkRunConcurrency` re-evaluates when the holding run finishes.

Covered by a unit test (`Test_checkJobsOfCurrentRunAttempt_RunLevelConcurrencyKeepsJobsBlocked` in `services/actions/job_emitter_test.go`) that sets up a Running holder attempt and a Blocked sibling attempt in the same concurrency group directly in the DB, calls `checkJobsOfCurrentRunAttempt`, and asserts the blocked job stays `Blocked`. Fails on master, passes with the fix.

---
This PR was written with the help of Claude Opus 4.7

### CODE REVIEW

File: tests/integration/actions_concurrency_test.go [hunk 1]
```diff
@@ -1219,6 +1220,71 @@ jobs:
...
+  job:
+    runs-on: ubuntu-latest
+    steps:
+      - run: echo 'schedule workflow'
+`
+		opts := getWorkflowCreateFileOptions(user2, repo.DefaultBranch, "create "+wfTreePath, wfFileContent)
+		createWorkflowFile(t, token, user2.Name, repo.Name, wfTreePath, opts)
+
+		// Push run picks up and holds the "schedule-concurrency" group while it runs.
+		_ = runner.fetchTask(t)
```
[the remark this answers is not in the document]
Comment (silverwind) [reply tests/integration/actions_concurrency_test.go]: `runner.fetchTask` already calls `require.NotNil(t, task, "failed to fetch a task")` internally (`tests/integration/actions_runner_test.go:97`), so the precondition is enforced. Discarding the return value with `_ =` is safe. — Claude (Opus 4.7)

File: services/actions/job_emitter.go [hunk 2]
```diff
@@ -228,6 +228,24 @@ func checkJobsOfCurrentRunAttempt(ctx context.Context, run *actions_model.Action
...
+		}
+		if has {
+			shouldBlock, err := shouldBlockRunByConcurrency(ctx, attempt)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("shouldBlockRunByConcurrency: %w", err)
+			}
+			if shouldBlock {
+				return jobs, nil, nil, nil
+			}
+		}
```
[the remark this answers is not in the document]
Comment (silverwind) [reply services/actions/job_emitter.go:247]: Falling through to the resolver is the correct behaviour for legacy attemptless runs (`RunAttemptID == 0`, pre-migration v331). Those runs don't use run-level concurrency at all, so there is no risk of premature unblocking, and bailing here would prevent the resolver from ever progressing their `Blocked` jobs through `needs` / job-level concurrency. — Claude (Opus 4.7)

File: tests/integration/actions_concurrency_test.go [hunk 3]
```diff
@@ -1219,6 +1220,71 @@ jobs:
...
+		fireSchedule()
+		require.NoError(t, queue.GetManager().FlushAll(t.Context(), 5*time.Second))
+
+		blockedRuns, err := db.Find[actions_model.ActionRun](t.Context(), actions_model.FindRunOptions{
+			RepoID:       repo.ID,
+			TriggerEvent: webhook_module.HookEventSchedule,
+			Status:       []actions_model.Status{actions_model.StatusBlocked},
+		})
+		require.NoError(t, err)
+		require.Len(t, blockedRuns, 1)
```
[the remark this answers is not in the document]
Comment (silverwind) [reply tests/integration/actions_concurrency_test.go]: Bumped to 15s in 1121db87af. `FlushAll` is already a synchronous drain, so an `Eventually` wrapper would be redundant. — Claude (Opus 4.7)

**Review by Zettat123 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
I believe it can be tested in a unit test.

`TestScheduleConcurrencyBlockedRunStaysBlocked` is extremely slow, and it is not clear about the test details.

**Comment by silverwind** [comment]
> TestScheduleConcurrencyBlockedRunStaysBlocked is extremely slow, and it is not clear about the test details.

Took 2.7s locally. Slow yes, but not extremely.

**Comment by wxiaoguang** [comment]
> > TestScheduleConcurrencyBlockedRunStaysBlocked is extremely slow, and it is not clear about the test details.
> 
> Took 2.7s locally. Slow yes, but not extremely.

These tests will be extremely slow in CI, the more added, the slower.

**Comment by wxiaoguang** [comment]
> > > TestScheduleConcurrencyBlockedRunStaysBlocked is extremely slow, and it is not clear about the test details.
> > 
> > 
> > Took 2.7s locally. Slow yes, but not extremely.
> 
> These tests will be extremely slow in CI, the more added, the slower.

Share some findings I have got about CI time: Operating a git repo via API/web is slow, due to it needs to execute the Gitea's git hook. It might be slower in CI due to limited resources.

To keep CI fast and optimize the speed: avoid unnecessary git repo operations via Gitea's API or web (avoid the Gitea's git hook) as much as possible

* Some tests can be clearly written in unit tests
* Some git repo operation can be done via `git fast-import`

I think it can save at least many minutes if `git fast-import` can be correctly used.

**Comment by silverwind** [comment]
Yes, with ~4 times slower CI, we are looking at 12s+, which is borderline.

**Comment by wxiaoguang** [comment]
Another concern is that the  integration tests are abused.

Actually, for a function level logic, it should be clearly tested in unit tests, including various edge cases.

Integration test should focus on "the whole thing overall works together", it's difficult to use it to cover edge cases, and usually it is not informative when writing a integration test for a speical case (a lot of unrelated code, maintenance burden)

**Comment by silverwind** [comment]
Yeah, what can be asserted in a unit test should be. I didn't tell Claude to write integration test, it decided itself. Likely a good pointer to add to AGENTS.md.

My attack points for fast CI are:

- Caching (https://github.com/go-gitea/gitea/pull/37462 + large enough cache space)
- Proper paralellization, ideally full test-level parallelization like playwright and vitest.

**Comment by silverwind** [comment]
@wxiaoguang good call — replaced the integration test with a unit test in `services/actions/job_emitter_test.go` (commit dd83ccf5a3). It sets up the Running holder + Blocked sibling attempts in the same concurrency group directly via `db.Insert` and calls `checkJobsOfCurrentRunAttempt` on the blocked run. ~0.3s vs ~3.7s for the integration version, no git-hook traffic. Verified it fails on master and passes with the fix.

---
This response was written with the help of Claude Opus 4.7

**Comment by wxiaoguang** [comment]
The new test looks good to me

**Review by lunny — APPROVED** [review]
